### PR TITLE
Implement ADR 0018

### LINKED ISSUE #1846: Performance of `bin` for many->many bin mappings
We previously optimized cases like this cases on the `SubbinSizes` helper class. Here is an example where this is not enough:

image

Bin into 50x50 output bins takes about `2 seconds`, even for virtually no events:
```python
sc.bin(a, erase=['detector_id'], edges=[theta_bins, wavelength_bins])
```

- Is there a way to ensure the optimization works better in this case?
- Note there are about 200 million mappings between bins, so performance hit is not unexpected.
- This issue might become much less relevant once we add multi-threading for reduction operations in `transform`. I think this should be the first thing to investigate, since it would also fix several other performance bottlenecks.

Note: A faster solution (for few events) is:
```python
sc.bin(d.bins.concatenate('detector_id'), edges=[theta_bins, wavelength_bins])
```

### LINKED ISSUE #3435: Is the mechanism for controlling whether `hist` replaces dimensions too indirect?
According to the docstring:
> When histogramming a dimension with an existing dimension-coord, the binning for
> the dimension is modified, i.e., the input and the output will have the same
> dimension labels.
>
> When histogramming by non-dimension-coords, the output will have new dimensions
> given by the names of these coordinates. These new dimensions replace the
> dimensions the input coordinates depend on.

In practice this means that:

- A prior use of `transform_coords` with or without the `rename_dims` option affects the outcome of a subsequent `hist`.
- It is possible to indirectly control which dimensions are to be removed, as shown in the example below, by renaming and/or flattening dimensions:

```python
import scipp as sc

table = sc.data.table_xyz(1000)
binned = table.bin(x=3, y=4)  # sizes {'x': 3, 'y': 4}
binned.rename_dims(y='z').hist(z=5)  # sizes {'x': 3, 'z': 5}
binned.flatten(to='z').hist(z=5)  # sizes {'z': 5}
```

The mechanism was introduced since it allows the algorithm to either add a new dimension, or replace an existing dimension. But is it too confusing when working with multi-dimensional data?

Would it suffice to improve the docstring (I am thinking of adding concrete examples on how to control the behavior), or do we need to think of something else?

Note that related functions such as `bin` are also affected.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #3435.
Fixes scipp/essreduce#47.
I also propose to close #1846. There are most likely still some remaining issues, but the original problem description has increasingly lower utility.

Implementing this change was simple at the beginning: Consider the first few commits, and you will see that adding the `dim` argument was very simple (since, e.g., the underlying `make_binned` already had an `erase` argument, which was almost exactly that). Apart from a few minor branching changes in behavior (https://github.com/scipp/scipp/pull/3501/commits/cd8de34e2a52b2c10631a7b21bb419d1f00756fa#diff-d64f8ed6d2f8ccfec8c613e469f7f4824afd55241e6775f1f433d7724dbb80cb) the tests just passed.

Now, for the not-so-nice part: Adding handling of the so far unsupported cases (mostly https://github.com/scipp/esssans/blob/a186d7eb7c57d2b367b097596b23b9ce425c0f35/src/ess/sans/i_of_q.py#L202-L259) and performance workarounds adds one more layer to the unfortunate stack of optimizations and handling of different cases in the Python and C++ implementations of binning and histogramming. I am not at all happy with this, and I am not confident that we will not find bugs in this (maybe not returning wrong results, but at least some optimized code branch raising exceptions in cases that should work). The entire thing is very hard to see through. We do have a good number of tests, and in many cases I saw existing tests flagging issues when I adding the new features and performance tweaks. But for now at least I am out of better ideas.

- [x] Works in ESSsans (can remove ~50 LOC)
- [x] Works in ESSdiffraction
- [x] Works in ESSreflectometry
- [x] Verify that performance is unchanged, compared to workarounds in ESS* packages.

To do:
- [ ] Update docstrings

### CODE REVIEW

**Review by jl-wynen — COMMENTED** [review]
Please also add an explanation in the user guide.

I did not manage to completely follow the new logic. But I also don't see a good way to make sure that all cases are covered by tests.

**Review by jl-wynen — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
@jl-wynen Given some mixed experiences in the past, I'd try to document this in the docstrings, not notebooks, unless you see a particular need for explaining this elsewhere?

**Comment by jl-wynen** [comment]
> @jl-wynen Given some mixed experiences in the past, I'd try to document this in the docstrings, not notebooks, unless you see a particular need for explaining this elsewhere?

Fine by me.

**Comment by SimonHeybrock** [comment]
@jl-wynen Could you please have a look at the latest commit? It only updates the docstring of `hist`, but I thought it is more efficient if you proof read it for clarity first, before I put equivalent descriptions on the other functions.

File: src/scipp/core/binning.py [hunk 1]
```diff
@@ -417,20 +417,30 @@ def hist(x, arg_dict=None, /, *, dim=None, **kwargs):
...
+    resulting in a new dimension in the output. In many cases this default yields the
+    desired behavior, there are two classes of exceptions where specifying `dim`
+    explicitly can be useful:
+
+    1. Given input data with an N-D coordinate, where N>1, we can use `dim` to restrict
+       the sum to a subset of M dimensions, resulting in an (N-M)-D "array" of histograms.
+    2. Given M-D input data with an N-D coordinate, where N<M, we can specify `dim` to
+       sum over, e.g., the remaining M-N dimensions while histogramming. This is often
+       equivalent to not specifying `dim` and a call to `sum` after histogramming but
+       is more memory efficient.
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/core/binning.py:433]: Please see added examples!

[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scipp/core/binning.py:433]: :+1: For the examples.

I understand that you don't want to duplicate the explanation. But can you at least add a link to `hist` from the other functions?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/scipp/core/binning.py:433]: I definitely want to duplicate the explanations, please see https://github.com/scipp/scipp/pull/3501#issuecomment-2309872650.

[the remark this answers is not in the document]
Comment (jl-wynen) [reply src/scipp/core/binning.py:433]: Oh, I did not see that comment. I think the docstring is good enough now.

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**